internal/selfupdate/download.go

```diff
@@ -74,33 +67,45 @@ func extractToFile(buf []byte, filename, target string, printf func(string, ...i
 		rd = file
 	}
 
-	err = os.Remove(target)
-	if os.IsNotExist(err) {
-		err = nil
-	}
+	// Write everything to a temp file
+	dir := filepath.Dir(target)
+	new, err := ioutil.TempFile(dir, "restic")
 	if err != nil {
-		return fmt.Errorf("unable to remove target file: %v", err)
+		return err
 	}
 
-	dest, err := os.OpenFile(target, os.O_CREATE|os.O_EXCL|os.O_WRONLY, mode)
+	n, err := io.Copy(new, rd)
 	if err != nil {
+		_ = new.Close()
+		_ = os.Remove(new.Name())
+		return err
+	}
+	if err = new.Sync(); err != nil {
+		return err
+	}
+	if err = new.Close(); err != nil {
 		return err
 	}
 
-	n, err := io.Copy(dest, rd)
-	if err != nil {
-		_ = dest.Close()
-		_ = os.Remove(dest.Name())
+	printf("saved %d bytes in %v\n", n, new.Name())
```

Comment:
That prints the tempfile name, which is rather confusing:
```
saved 18472960 bytes in /restic/restic/restic3463974662
```
 Please print the output using the final filename and print it only once the update is complete.